### PR TITLE
[Nu-1407] Improve validation in numeric toNumber helper method

### PR DESCRIPTION
**Pull request by DeamonDev**
## Checklist before merge
- [x] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [x] Code is cleaned from temporary changes and commented out lines
- [x] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala [hunk 1]
```diff
@@ -57,7 +59,27 @@ trait numeric extends MathUtils with HideToString {
   @Documentation(description = "Parse string to number")
   @GenericType(typingFunction = classOf[ToNumberTypingFunction])
   def toNumber(@ParamName("stringOrNumber") stringOrNumber: Any): java.lang.Number = stringOrNumber match {
-    case s: CharSequence     => new java.math.BigDecimal(s.toString)
+    case s: CharSequence =>
+      val ss                                       = s.toString
+      val tryByte: Try[java.lang.Byte]             = Try(java.lang.Byte.parseByte(ss))
```
Comment (arkadius) [inline utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala:64]: I think that the better approach will be behaving as java behaves when someone declares literal number - I mean skipping byte, short and float cases. I think that these cases are only for some hard computations when you need to save memory. In most business cases, they introduce more noise than it benefits.

File: utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala [hunk 2]
```diff
@@ -57,7 +59,27 @@ trait numeric extends MathUtils with HideToString {
...
+      val tryShort: Try[java.lang.Short]           = Try(java.lang.Short.parseShort(ss))
+      val tryInteger: Try[java.lang.Integer]       = Try(java.lang.Integer.parseInt(ss))
+      val tryLong: Try[java.lang.Long]             = Try(java.lang.Long.parseLong(ss))
+      val tryFloat: Try[java.lang.Float]           = Try(java.lang.Float.parseFloat(ss))
+      val tryDouble: Try[java.lang.Double]         = Try(java.lang.Double.parseDouble(ss))
+      val tryBigDecimal: Try[java.math.BigDecimal] = Try(new java.math.BigDecimal(ss))
+      val tryBigInteger: Try[java.math.BigInteger] = Try(new java.math.BigInteger(ss))
+
+      val tries: List[Try[java.lang.Number]] =
```
Comment (arkadius) [inline utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala:73]: Let's skip local variables above and keep everything in the inlined list. Thanks to that we avoid the mistake that this list misses some case from above.

File: utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala [hunk 3]
```diff
@@ -57,7 +59,27 @@ trait numeric extends MathUtils with HideToString {
...
+        List(tryByte, tryShort, tryInteger, tryLong, tryFloat, tryDouble, tryBigDecimal, tryBigInteger)
+
+      val firstSuccess: Option[Number] = tries
+        .collectFirst { case Success(value) =>
+          value
+        }
+        .orElse(Some(new java.math.BigDecimal(ss)))
```
Comment (arkadius) [inline utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala:80]: `orElse(Some(...).get` is the same as `getOrElse(...)`

File: utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/CollectionUtilsSpec.scala [hunk 4]
```diff
@@ -189,9 +189,9 @@ class CollectionUtilsSpec extends AnyFunSuite with BaseSpelSpec with Matchers {
     evaluateType("#COLLECTION.sum({#int, #long})", variables) shouldBe "Number".valid
   }
 
-  test("sum should fall back to Double on unknown Number type") {
```
Comment (arkadius) [inline utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/CollectionUtilsSpec.scala:192]: Please add tests:
1. doing the same but using simple plus operator
2. showing that we behave different than before in the runtime

Comment (DeamonDev) [reply utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/CollectionUtilsSpec.scala:192]: I've added tests

File: utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala (hunk 3 above)
Comment (DeamonDev) [reply utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala]: Done

File: utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala (hunk 2 above)
Comment (DeamonDev) [reply utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala]: Done

File: utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/CollectionUtilsSpec.scala (hunk 4 above)
Comment (arkadius) [reply utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/CollectionUtilsSpec.scala:192]: What about tests mentioned in 2. point?

File: utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/NumericUtilsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatest.matchers.should.Matchers
+import cats.implicits._
+
+import java.math.BigInteger
+
+class NumericUtilsSpec extends AnyFunSuite with BaseSpelSpec with Matchers {
+
+  test("numeric") {
```
Comment (arkadius) [inline utils/default-helpers/src/test/scala/pl/touk/nussknacker/engine/util/functions/NumericUtilsSpec.scala:11]: I would call it `test("string to number conversion should return narrow numeric types in the runtime")` or sth similar. Thanks to that we would know in the future, why we introduced this change.

File: utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala [hunk 6]
```diff
@@ -57,7 +59,23 @@ trait numeric extends MathUtils with HideToString {
   @Documentation(description = "Parse string to number")
   @GenericType(typingFunction = classOf[ToNumberTypingFunction])
   def toNumber(@ParamName("stringOrNumber") stringOrNumber: Any): java.lang.Number = stringOrNumber match {
-    case s: CharSequence     => new java.math.BigDecimal(s.toString)
+    case s: CharSequence =>
+      val ss = s.toString
+
+      val tries: List[Try[java.lang.Number]] = List(
```
Comment (arkadius) [inline utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/numeric.scala:65]: I would add a comment that we pick the narrowest type as possible to reduce the amount of memory and computations overheads ,

**Review by arkadius — APPROVED** [review]
LGTM, I added some small comments imrovements that could help us to understand why we introduced changes in the `toNumber` runtime.

**Pull request merged**